sphinx/tutorial/rose/furthertopics/upgrading-macro-development.rst

```diff
@@ -0,0 +1,568 @@
+
+       BEFORE_TAG = "0.1"
+       AFTER_TAG = "0.2"
+
+       def upgrade(self, config, meta_config=None):
+           """Upgrade the boat!"""
+           # Some code doing something to config goes here.
+           return config, self.reports
+
+This is already a functional upgrade macro - although it won't do anything.
+
+.. note::
+
+   The name of the class (``MyFirstUpgradeMacro``) doesn't need to
+   be related to the versions - the only identifiers that matter are the
+   ``BEFORE_TAG`` and the ``AFTER_TAG``.
+
+We need to get the macro to do the following:
+
+* add the option ``namelist:materials=misc_branches``
+* add the option ``namelist:materials=outrigger_tree_trunks``
+* add the option ``namelist:materials=paddling_branches``
+* remove the option ``namelist:materials=paddling_twigs``
+
+We can use the nice API provided to express this in Python code - replace
+the ``# Some code doing something...`` line with:
+
+.. code-block:: python
+
+        self.add_setting(config, ["namelist:materials", "misc_branches"], "4")
+        self.add_setting(
+                 config, ["namelist:materials", "outrigger_tree_trunks"], "2")
+        self.add_setting(
+                 config, ["namelist:materials", "paddling_branches"], "1")
+        self.remove_setting(config, ["namelist:materials", "paddling_twigs"])
+
+This changes the app configuration (``config``) in the way we want, and
+(behind the scenes) adds some things to the ``self.reports`` list
+mentioned in the ``return config, self.reports`` line.
+
+.. note::
+
+   When we add options like ``misc_branches``, we need values to add them
+   with - what they are will often be a matter of judgement.
+
+However, even if they appear to be numeric inputs, values should always be
+entered as strings (``"1"`` rather than ``1``).
+
+Customising the Output
+^^^^^^^^^^^^^^^^^^^^^^
+
+The methods ``self.add_setting`` and ``self.remove_setting`` will provide
+a default message to the user about the change (e.g.
+``"Added X with value Y"``), but you can customise them to add your own
+by passing an info 'keyword argument' like this:
+
+.. code-block:: python
+
+           self.add_setting(
+                    config, ["namelist:materials", "outrigger_tree_trunks"], "2",
+                    info="This makes it into a trimaran!")
+
+If you want to, try adding your own messages.
+
+Running rose app-upgrade
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Our upgrade macro will now work - change directory to the application
+directory and run:
+
+.. code-block:: bash
+
+   rose app-upgrade --meta-path=~/rose-meta/
+
+``--meta-path`` equals the path to the ``rose-meta/`` directory you
+created - as this path isn't configured in the site/user configuration,
+we need to set it manually. This won't normally be the case for users,
+if the metadata is centrally managed.
+
+This should display some information about the current and available
+versions - see the help by running ``rose help app-upgrade``.
```

Comment:
The command on line 317 isn't doing the actual upgrade but this is only stated explicitly here and not clarified immediately after, so there is the potential for confusion given the introductory statement on line 312 - I recommend moving lines 324-325 to directly follow the command declaration to resolve.